src/Microsoft.OData.Core/ODataInnerError.cs

```diff
@@ -30,41 +31,80 @@ public ODataInnerError()
+            : this(exception, null)
         {
-            ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
-            this.Message = exception.Message ?? string.Empty;
-            this.TypeName = exception.GetType().FullName;
-            this.StackTrace = exception.StackTrace;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:Microsoft.OData.ODataInnerError" /> class with exception object.
+        /// </summary>
+        /// <param name="exception">The <see cref="T:System.Exception" /> used to create the inner error.</param>
+        /// <param name="nestedInnerErrorName">The name of the field for the nested inner error object inside inner error.</param>
+        public ODataInnerError(Exception exception, string nestedInnerErrorName)
+        {
+            ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
             if (exception.InnerException != null)
             {
-                this.InnerError = new ODataInnerError(exception.InnerException);
+                this.InnerError = new ODataInnerError(exception.InnerException, nestedInnerErrorName);
             }
+
+            Properties = new Dictionary<string, ODataValue>();
+            Properties.Add(JsonConstants.ODataErrorInnerErrorMessageName, exception.Message.ToODataValue() ?? string.Empty.ToODataValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorTypeNameName, exception.GetType().FullName.ToODataValue() ?? string.Empty.ToODataValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorStackTraceName, exception.StackTrace.ToODataValue() ?? string.Empty.ToODataValue());
+        }
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="T:Microsoft.OData.ODataInnerError" /> class with a dictionary of property names and corresponding ODataValues.
+        /// </summary>
+        /// <param name="properties">Dictionary of string keys with ODataValue as value. Key string indicates the property name where as the value of the property is encapsulated in ODataValue.</param>
+        /// <param name="nestedInnerErrorName">The name of the field for the nested inner error object inside inner error.</param>
+        /// <param name="nestedObject">Nested inner error object.</param>
+        public ODataInnerError(IDictionary<string, ODataValue> properties, string nestedInnerErrorName, ODataInnerError nestedObject)
+        {
+            Properties = properties;
+            this.NestedObjectName = nestedInnerErrorName;
+            InnerError = nestedObject;
+        }
+
+        /// <summary>
+        /// Nested object's name
+        /// </summary>
+        internal string NestedObjectName;
+
+        /// <summary>
+        /// Properties to be written for the inner error.
+        /// </summary>
+        public IDictionary<string, ODataValue> Properties
+        {
+            get;
+            set;

```

Comment:
>set; [](start = 12, length = 4)

I don't think it's ok to public set this "Properties", because you set it in the constrctor